sympy/holonomic/holonomic.py

```diff
@@ -518,19 +558,126 @@ def integrate(self, *args):
+                x0 = self.x0
+                a = limits[1]
+                b = limits[2]
+
+        else:
+            x0 = self.x0
+            a = self.x0
+            b = self.x
+
+        if x0 == a:
             y0 = [S(0)]
             y0 += self.y0
-            return HolonomicFunction(self.annihilator * D, self.x, self.x0, y0)
+
+        # use evalf to get the values at `a`
+        else:
+            y0 = [S(0)]
+            tempy0 = self.evalf(a, derivatives=True)
+            y0 += tempy0
+
+        # if the upper limit is `x`, the answer will be a function
+        if b == self.x:
+            return HolonomicFunction(self.annihilator * D, self.x, a, y0)
+
+        # if the upper limits is a Number, a numerical value will be returned
+        elif S(b).is_Number:
+            return HolonomicFunction(self.annihilator * D, self.x, a, y0).evalf(b)
 
         return HolonomicFunction(self.annihilator * D, self.x)
 
+    def diff(self, *args):
+        """
+        Differentiation of the given Holonomic function.
+
+        Examples
+        ========
+
+        >>> from sympy.holonomic.holonomic import HolonomicFunction, DifferentialOperators
+        >>> from sympy.polys.domains import ZZ, QQ
+        >>> from sympy import symbols
+        >>> x = symbols('x')
+        >>> R, Dx = DifferentialOperators(ZZ.old_poly_ring(x),'Dx')
+
+        # derivative of sin(x)
+        >>> HolonomicFunction(Dx**2 + 1, x, 0, [0, 1]).diff().to_sympy()
+        cos(x)
+
+        # derivative of e^2*x
+        >>> HolonomicFunction(Dx - 2, x, 0, 1).diff().to_sympy()
+        2*exp(2*x)
+
+        See Also
+        =======
+
+        .integrate()
+        """
+
+        if args:
+            if args[0] != self.x:
+                return S(0)
+            elif len(args) == 2:
+                sol = self
+                for i in range(args[1]):
+                    sol = sol.diff(args[0])
+                return sol
+
+        ann = self.annihilator
+        dx = ann.parent.derivative_operator
+
+        # if the function is constant.
+        if ann.listofpoly[0] == ann.parent.base.zero and ann.order == 1:
+            return S(0)
+
+        # if the coefficient of y in the differential equation is zero.
+        # a shifting is done to compute the answer in this case.
+        elif ann.listofpoly[0] == ann.parent.base.zero:
+
+            sol = DifferentialOperator(ann.listofpoly[1:], ann.parent)
+
+            if self._have_init_cond:
+                return HolonomicFunction(sol, self.x, self.x0, self.y0[1:])
+
+            else:
+                return HolonomicFunction(sol, self.x)
+
+        # the general algorithm
+        R = ann.parent.base
+        K = R.get_field()
+
+        seq_dmf = [K.new(i.rep) for i in ann.listofpoly]
+
+        # -y = a1*y'/a0 + a2*y''/a0 ... + an*y^n/a0
+        rhs = [i / seq_dmf[0] for i in seq_dmf[1:]]
+        rhs.insert(0, K.zero)
+
+        # differentiate both lhs and rhs
+        sol = _derivate_diff_eq(rhs)
```

Comment:
This function is defined in master, but I think `derivate` is the wrong word. Did you mean to use `differentiate` or `derive`?